lib/internal/modules/cjs/loader.js

```diff
@@ -239,16 +241,22 @@ Module._debug = deprecate(debug, 'Module._debug is deprecated.', 'DEP0077');
 //   -> a
 //   -> a.<ext>
 //   -> a/index.<ext>
+/**
+ *
+ * @param {string} jsonPath
+ * @returns {string|undefined}
+ */
+function readJsonDefaultStrategy(jsonPath) {
+  return internalModuleReadJSON(path.toNamespacedPath(jsonPath));
+}
 
-const packageJsonCache = new SafeMap();
-
-function readPackage(requestPath) {
+function readPackage(requestPath, readJsonStrategy = readJsonDefaultStrategy) {
   const jsonPath = path.resolve(requestPath, 'package.json');
 
   const existing = packageJsonCache.get(jsonPath);
   if (existing !== undefined) return existing;
 
-  const json = internalModuleReadJSON(path.toNamespacedPath(jsonPath));
+  const json = readJsonStrategy(jsonPath);
```

Comment:
This looks like we're using different functions to populate the same cache key. Is there any danger of race conditions where one strategy writes a key and another then silently gets the value?

Reply:
Goog point! 

> Is there any danger of race conditions where one strategy writes a key and another then silently gets the value?

The worst scenario is the following. CJS reads invalid package.json and puts in the cache `{}`. ESM module then reads the `{}` and tries to work with it, although ESM should throw error in this case.

I believe it would be better to use `internalModuleReadJSON` for both ESM and CJS. First, It allows to use the same cache. Second, `internalModuleReadJSON` is resilient to failure.
We would need to sacrifice nice error reporting for ESM. But we could explicitly describe this behavior (the case when package.json is invalid) in docs and it won't be a problem.

 @guybedford what do you think?

Reply:
Could we make it always return the same `{}` and expose it as `INVALID_PACKAGE_JSON`? That way ESM could check for strict equality and maybe still throw..?

Reply:
Yes this does sound like a problem. Thanks @jkrems for spotting this.

This does sound like a serious inconsistency with sharing the caches, with the only solutions that aren't unsound being:

1. Change the ESM resolver to pass on invalid JSON without these fields, like this does for the CJS resolver.
2. Fixup `internalModuleReadJSON` so that it's parser actually detects invalid JSON state.

I suspect (2) is probably the best approach moving forward here to be honest.

@trevnorris we may need to create a new issue here to track the fact that `internalModuleReadJSON` will not throw JSON parsing errors for invalid syntax JSON files without the search keywords in them. Do you think a fix to ensure invalid JSON parser state is comprehensively detected to return `undefined` would be straightforward.

@shackijj if you're proficient with C++ it might be worth looking into as well.

Reply:
@bnoordhuis we may need to create a new issue here to track the fact that `internalModuleReadJSON` will not throw JSON parsing errors for invalid syntax JSON files without the search keywords in them. Do you think a fix to ensure invalid JSON parser state is comprehensively detected to return `undefined` would be straightforward?

Apologies @trevnorris for copying you in by mistake :)

Reply:
@guybedford CJS loader detects invalid JSON state [here](https://github.com/nodejs/node/blob/8a6fab02adab2de05f6e864847f96b0924be0840/lib/internal/modules/cjs/loader.js#L263). I think that parsing JSON in `internalModuleReadJSON` might lead to performance issues unless there is a way to parse string there (in `internalModuleReadJSON`) and get a reference to a resulting (parsed) object in JS (i'm not sure if it's possible).

Alternatively, we might create a cache that will store only package.json strings (not parsed objects as it's now). ESM and CJS could parse the string the way they do now and have their own caches that will store parsed objects.

![image](https://user-images.githubusercontent.com/9201657/81545819-22607600-9382-11ea-97d8-8c3ab3f7f601.png)

Reply:
@shackijj the specific issue is a package.json file like `{ "invalid" }` is detected by `internalModuleReadJSON` as not containing properties `"main"`, `"exports"`, `"type"` or `"name"` so that function returns `"{}"` - the empty object string, instead of invalid. So CommonJS isn't currently throwing on these files.

`internalModuleReadJSON` is already a rudimentary JSON parser. It doesn't seem like it would be a big stretch to give it the ability itself in the parsing loop to detect invalid syntax and explicitly return `undefined` in these cases to indicate the error.

Reply:
>  It doesn't seem like it would be a big stretch to give it the ability itself in the parsing loop to detect invalid syntax and explicitly return undefined in these cases to indicate the error.

It's possible, but it will significantly change the behaviour of CJS (seems like breaking a change to me). Second, I think, that we will need to reimplement JSONParse in `internalModuleReadJSON`.

Reply:
It feels like an extreme edge case to me that there's a `package.json` that both has syntax errors _and_ doesn't include any of the relevant metadata fields. Should we just accept the CJS behavior for ESM here? The fallout seems extremely limited.

Reply:
Yes it is an edge case, but the reason we have a resolver specification that is public is because we aim to define all the edge cases comprehensively so that other projects can implement resolver logic that matches Node.js exactly.

Adding a line to the resolver specification like "if the JSON does not contain the strings main, name, type or exports then do not throw any invalid JSON syntax errors" seems like a very bad idea to me (within the edge case of course) as it's highly inconsistent. 

@shackjj I agree the CJS case would be breaking to change, but in theory it was also breaking when it was first added. There is actually another way we could do this actually. Instead of returning `emptyObjectString` in the `internalModuleReadJSON` function, we could instead return an object like `{ source: string | undefined, containsKeys: boolean }`. This would then split the function down into three cases:

1. `source: string, containsKeys: true`: String for JSON source when it has the keys (same between esm and cjs loaders)
2. `source: undefined, containsKeys: any`: Missing file, same between esm and cjs loaders
3. `source: string, containsKeys: false`: For CJS source it just treats it as am empty object. For ESM it would still parse the source to return any parse error if necessary.

Reply:
> Adding a line to the resolver specification like "if the JSON does not contain the strings main, name, type or exports then do not throw any invalid JSON syntax errors" seems like a very bad idea to me (within the edge case of course) as it's highly inconsistent.

Agree

> Instead of returning string | undefined from the internalModuleReadJSON function, we could instead return an object like { source: string | undefined, containsKeys: boolean }

Good idea! I need to read about C++/JS communication in node because I haven't worked with v8 C++ API before. Will try to submit a patch during this week.

@bnoordhuis Could you please share you thoughts on @guybedford's suggestion?

Reply:
Sounds good - if it helps, you should be able to create an object something like https://github.com/nodejs/node/blob/master/src/node_process_object.cc#L93 and set properties something like https://github.com/nodejs/node/blob/master/src/node_process_object.cc#L157. I'm terrible at this stuff so don't quote me :)